Source/Dafny/AST/DafnyAst.cs

```diff
@@ -1062,25 +1062,14 @@ public enum AutoInitInfo { MaybeEmpty, Nonempty, CompilableValue }
     }
 
     /// <summary>
-    /// Returns true if it's possible to check at run-time that a value has this type
+    /// Returns true if it's surely possible to check at run-time that a value has this type
     /// </summary>
     /// <returns>A boolean indicating if the type is compilable, meaning run-time checkable.</returns>
-    public bool IsCompilable() {
-      // It might not be possible to do so if the constraint of this type uses ghost predicates
-      if (this.AsSubsetType is { IsConstraintCompilable: false }) {
-        return false;
-      } else if (this.NormalizeExpandKeepConstraints() is TypeProxy) {
-        // If we are still in the early resolution phase and no type has been inferred,
-        // then we cannot say for sure that the type is compilable.
-        return false;
-      } else {
-        return true;
-      }
-    }
+    public abstract bool DoesNotContainGhostConstraints();
```

Comment:
Consider changing this to `ContainsGhostConstraints` to remove the negation, and also the double negation in `!v.Type.DoesNotContainGhostConstraints()` in `ScopePushBoundVarsAssumingCompilable`. I'm guessing you added the negation here because of the special `return false` case in 

```
    public override bool DoesNotContainGhostConstraints() {
      if (this.AsSubsetType is { IsConstraintCompilable: false }) {
        return false;
      } else if (this.NormalizeExpandKeepConstraints() is TypeProxy) {
        // If we are still in the early resolution phase and no type has been inferred,
        // then we cannot say for sure that the type is compilable.
        return false;
      } else {
        return true;
      }
    }
```

Maybe you dislike returning `false` when really you're not sure. I think returning `true` there if the method was called `ContainsGhostConstraints` would not be worse. You can also consider returning `null` there but I think this is fine.

Reply:
I negated the method, renamed it `MightContainGhostConstraints`.